core/perfutil/perfutil.go

```diff
@@ -0,0 +1,252 @@
+	if enableTrace {
+		to := allTransactionTraces[uuid]
+		if to == nil {
+			fmt.Printf("[GetPT] Transaction trace nil. \n")
+			return to
+		}
+		for {
+			if to.next == nil {
+				break
+			}
+			to = to.next
+		}
+		return to
+	}
+	return nil
+}
+
+//create the hash table entry and the first trace
+func CreatePTHdr(uuid string, name string, now int64) {
+	if enableTrace {
+		perfUuid = uuid
+		if allTransactionTraces[uuid] != nil {
+			fmt.Printf("Transaction trace exists\n")
+			return
+		}
+
+		if now == 0 {
+			now = time.Now().UnixNano()
+		}
+		allTransactionTraces[uuid] = &traceOperation{ name: name, start: now }
+	}
+}
+
+//create trace from a divide (gRPC, channel, etc)
+func CreatePTOPFromCC(tt *traceOperation, name string, since int64) *traceOperation {
+	if tt == nil {
+		return nil
+	}
+
+	to := CreatePTOP(tt, name, since)
+	//Message (e.g., ChaincodeMessage) just computes time taken across a divide (grpc, channel, etc)
+	//for this, if we got the message, it is a success
+	EndCurrPTOP(to, true)
+	return to
+}
+
+//create trace
+func CreatePTOP(tt *traceOperation, name string, now int64) *traceOperation {
+	if tt == nil {
+		fmt.Printf("[CreatePTOP] ... Transaction trace nil: invalid\n")
+		return nil
+	}
+	if name == "" {
+		panic("blank PTOP name")
+	}
+
+	if now == 0 {
+		now = time.Now().UnixNano()
+	}
+
+	// sanity check current entry
+	if tt.start == 0 || tt.end == 0 {
+		fmt.Printf("ERROR create called for [%s] when current [%s] not set correctly (%d, %d)\n", name, tt.name, tt.start, tt.end)
+	}
+
+	tt.next =  &traceOperation{ name: name, start: now }
+	return tt.next
+}
+
+// end the current trace
+func EndCurrPTOP(tt *traceOperation, res bool) {
+	if tt == nil {
+		fmt.Printf("ERROR endCurrent called when create not called\n")
+		return
+	}
+
+	if tt.start == 0 {
+		fmt.Printf("ERROR endCurrent called when start not set for [%s]\n", tt.name)
+		return
+	}
+	tt.end = time.Now().UnixNano()
+	tt.succ = res
+}
+
+// PerfTraceHandler handles performance trace instrumentation request
+// input:
+//    uuid: hash table id
+//    name: name of this entry
+//    now:  timestamp
+//    res:  result of the corresponding API
+//    action: type of actions
+//            Create - create hash table entry and start a link list entry with the first entry having the name provided
+//			do not care: res
+//            CreatePTOP - close current link list entry and create a new link list entry with the name provided
+//			do not care: res
+//            EndPTOP - close current link list entry and end current link list
+//			do not care: name and now
+//            CreatePTOPFrom - close current link list entry, create a link list entry with the name and timestamps (now) provided,
+//                             and start a new one from Now() with the same trace name as the newly created entry with preface "Since_"
+//			do not care: res
+func PerfTraceHandler(uuid string, name string, now int64, res bool, action string) {
```

Comment:
I think this API entry is too complex.  I suggest providing multiple functions, maybe even one per action type.

Most of the time `PerfTraceHandler` is being invoked with the same set of arguments - `now = 0, res = true, action = "CreatePTOP"` (often also `uuid = GetPerfUuid()`).

Lastly, `perfutil.PerfTraceHandler()` stutters.  I suggest dropping `Perf` from all function names, and shortening them.  E.g. `perfutil.Trace("foofunc")`.